src/lavinmq/auth/oauth_user.cr

```diff
@@ -0,0 +1,51 @@
+require "./base_user"
+
+module LavinMQ
+  module Auth
+    class OAuthUser < BaseUser
+      getter name : String
+      getter tags : Array(Tag)
+      getter permissions : Hash(String, Permissions)
+
+      @token_updated = Channel(Nil).new
+
+      def initialize(@name : String, @tags : Array(Tag), @permissions : Hash(String, Permissions),
+                     @expires_at : Time, @authenticator : TokenVerifier)
+      end
+
+      def token_lifetime : Time::Span
+        @expires_at - RoughTime.utc
+      end
+
+      def on_expiration(&block)
+        spawn(name: "OAuthUser#on_expiration") do
+          loop do
+            select
+            when @token_updated.receive
+              next
+            when timeout(token_lifetime)
+              block.call
+              break
+            end
+          end
+        end
+      end
```

Comment:
This loop could be a problem. If the user is deleted nothing kills this fiber and stops this loop and after a while `token_lifetime` will become a negative value making this loop very fast. 

Could we do like 


```crystal
      def on_expiration(&block)
        spawn(name: "OAuthUser#on_expiration") do
          loop do
            lifetime = token_lifetime
            break if lifetime.negative? 
            select
            when @token_updated.receive
              next
            when timeout(lifetime)
              block.call
              break
            end
          end
        end
      end
```

or is there a better solution? @carlhoerberg ideas?

Reply:
closing the channel to stop the fiber feel like a good approach, @kickster97 can you have a look at this? I think we should address this before merge.

Reply:
I pushed a solution [a1d5e3b](https://github.com/cloudamqp/lavinmq/pull/1172/commits/a1d5e3b6af0e3da0f1347676dd807131ae1b2381) for closing the loop/fiber when a connection is ended. 

But I don't really understand the "fast looping", i could not reproduce this.. I think a negative value will just trigger time out? I'm not against putting the negative value check in the loop, i just don't understand it :)  If you want we can have a call and discuss

Reply:
Hehe yeah I was a bit confused about that :)  
No, i still think we want to have a cleanup for the user, if the client disconnects before the token expires its good that we can clean up the fiber.. otherwise we could get a fiber leak if a use has a high connection churn with with a token that has a long ttl